src/coreclr/vm/jitinterface.cpp

```diff
@@ -9849,6 +9846,125 @@ CorInfoUnmanagedCallConv CEEInfo::getUnmanagedCallConv(CORINFO_METHOD_HANDLE met
     return result;
 }
 
+/*********************************************************************/
+
+    // return the entry point calling convention for any of the following
+    // - a P/Invoke
+    // - a method marked with UnmanagedCallersOnly 
+    // - a function pointer with the CORINFO_CALLCONV_UNMANAGED calling convention.
+CorInfoCallConvExtension CEEInfo::getEntryPointCallConv(CORINFO_METHOD_HANDLE method, CORINFO_SIG_INFO* callSiteSig)
+{
+    CONTRACTL {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_PREEMPTIVE;
+    } CONTRACTL_END;
+
+    CorInfoCallConvExtension callConv = CorInfoCallConvExtension::Managed;
+
+    JIT_TO_EE_TRANSITION();
+
+    _ASSERTE(method != nullptr || callSiteSig != nullptr);
+
+    if (method)
+    {
+        MethodDesc* pMD = GetMethod(method);
+        if(pMD->IsNDirect())
+        {
+            EX_TRY
+            {
+                PInvokeStaticSigInfo sigInfo(pMD, PInvokeStaticSigInfo::NO_THROW_ON_ERROR);
+
+                switch (sigInfo.GetCallConv()) {
+                    case pmCallConvCdecl:
+                        callConv = CorInfoCallConvExtension::C;
+                        break;
+                    case pmCallConvStdcall:
+                        callConv = CorInfoCallConvExtension::Stdcall;
+                        break;
+                    case pmCallConvThiscall:
+                        callConv = CorInfoCallConvExtension::Thiscall;
+                        break;
+                    default:
+                        callConv = CorInfoCallConvExtension::Managed;
+                        break;
+                }
+            }
+            EX_CATCH
+            {
+                callConv = CorInfoCallConvExtension::Managed;
+            }
+            EX_END_CATCH(SwallowAllExceptions)
+        }
+#if !defined(TARGET_X86)
```

Comment:
This ifdef should not be needed.

Reply:
The ifdef is needed since for x86 we still handle all of the calling convention differences in the entry point in the Reverse P/Invoke thunk. If/when the JIT can emit the entry point of a function with a calling convention that matches the native calling conventions, then we can remove this ifdef.

Reply:
For PInvoke calls, there is `CORINFO_FLG_PINVOKE` flag.
For unmanaged calli, we should have just one `UNMANAGED` call conv in `CORINFO_SIG_INFO` and get rid of the other unmanaged calling conventions from there. This new API that you are introducing would be the one and only proper way to get the unmanaged calling convention details.
For reverse PInvoke, there is `CORJIT_FLAG_REVERSE_PINVOKE` flag.

Reply:
Agree. My point was that if you remove this ifdef, everything is still going to work correctly today, we will just have a bit of dynamically unreachable code on x86 for now.

Reply:
It won't work today for UnmanagedCallersOnly methods marked with Thiscall. Those would get the transformation in the JIT for the return buffer (that's not ifdef'd out for x86), but we currently handle Thiscall specially in the reverse p/invoke thunk.

Reply:
I agree that it may be a problem today. It should not be a problem if you do the other part of the change to make the JIT call this method for `CORJIT_FLAG_REVERSE_PINVOKE` methods only.

We never pass `CORJIT_FLAG_REVERSE_PINVOKE` to the JIT on x86, so the JIT should not ever call this method for UnmanagedCallersOnly on x86.